### ISSUE TITLE
Add new scriptlet — 'trusted-json-set'

### ISSUE THREAD

**Issue by contribucious** [AdguardTeam/Scriptlets#308]
### Hello,

Basically, in the continuity of the `json-prune` scriptlet (and the `$jsonprune` modifier, allowing more advanced queries [_currently_](https://github.com/AdguardTeam/Scriptlets/issues/183#issuecomment-1336487216)), it can be pretty neat and logical to have a <code><em><strong>trusted-</strong></em>json-set</code> scriptlet (and a `$jsonset` modifier), I would say. To either define new or modify existing JSON properties, instead of pruning them.

Some websites / web apps (TF1, M6 too nowadays, etc.) require certain (ad-related) JSON properties to be both present and boolean.
↪️ _Currently fixed using JS rules but hey … let's improve/simplify that! :wink:_

###### P.S. In a discussion I had a while ago with @ameshkov, there was talk of a possible untrusted version limited to true/false/etc. (i.e. a basic `json-set` scriptlet, thus following the same pattern as `set-constant` and `trusted-set-constant`), but probably still too risky though, I would say.

&nbsp;
### Related discussions:
- https://github.com/AdguardTeam/CoreLibs/issues/1447
- https://github.com/AdguardTeam/Scriptlets/issues/106

**Comment by contribucious** [comment]
## 🆕 Added in uBlockOrigin 1.64.0

- [Add json-edit suite of scriptlets; extend replace= option](https://github.com/gorhill/uBlock/commit/b18daa53aa)

<sub>⤷ https://github.com/gorhill/uBlock/releases/tag/1.64.0</sub>
<sub>⤷ https://www.reddit.com/r/uBlockOrigin/comments/1km1elg/ublock_origin_ubo_164_announcement_thread/</sub>

###### 💡 See also: [`Add tool to evaluate uBO-flavored JSONpath expressions`](https://github.com/gorhill/uBlock/commit/4888ec6d35739f43933f810398c1666c9d249a66) (+ [`Minor change`](https://github.com/gorhill/uBlock/commit/cdd46c5c7bd95b8ae07110626352ed1e975b9b1c)) — [FULL FILE HISTORY](https://github.com/gorhill/uBlock/commits/master/tools/jsonpath-tool.html)
<br>

_Thank you very much, @gorhill!_

**Comment by contribucious** [comment]
# Related

See also @gorhill's new commit: [`Add json-edit-related scriptlets`](https://github.com/gorhill/uBlock/commit/87e0434c90eaff27a5b6eabf6426537e2f7a1643) 🥳

_Thanks again to him!_

### Excited to see all this implemented in AdGuard products one day.

**Comment by contribucious** [comment]
👀 Issue can probably be closed as should now be tracked here instead I guess: https://github.com/AdguardTeam/Scriptlets/issues/523 and https://github.com/AdguardTeam/Scriptlets/issues/522.

**Comment by contribucious** [comment]
Hello @AdamWr,

Just posted these two commit comments on `src/scriptlets/trusted-json-set.ts` now: 😊
- https://github.com/AdguardTeam/Scriptlets/commit/fa5555bfb34d85fef31697d494c9694bdb74f605#r179541230
- https://github.com/AdguardTeam/Scriptlets/commit/fa5555bfb34d85fef31697d494c9694bdb74f605#r179541252

###### Note: These direct links to the comments do not work in this case because the file in question is not loaded (`"Large diffs are not rendered by default."`) unless you click `Load Diff`.

Thanks again! 🙂🎉

**Comment by contribucious** [comment]
I indeed spotted this parameter and thought about hijacking its initial intent (which is to log before and after a change), but that's more a workaround than a proper lightweight solution (like `json-prune` has), especially because EVERYTHING is printed twice (before/original and after/modified, even when identical). So, something more lightweight in both its required parameters and its output.

Also, something potentially very large like `example.org#%#//scriptlet('trusted-json-set', 'JSON.stringify')` to log all `JSON.stringify` calls, similar to the logging of all `JSON.parse` calls available via the `json-prune` scriptlet, is very problematic for example when everything is printed twice.

_What are your thoughts about that? Also, about the listed alternatives?_ 🙂

**Comment by contribucious** [comment]
In case you're up for it, should I open a dedicated issue? 👍

#### If so, which of the following? 🙂

- Improve 'trusted-json-set' — add logging support
- Add new scriptlet — 'json-log'
- Add new scriptlet — 'log-json'

###### For a new scriptlet, because it's only logging, no need to be a trusted one I would say.

**Comment by AdamWr** [comment]
You can open whatever issue you want :)

I just don't think that adding new scriptlet is necessary when it's possible to achieve similar result with current one + search/filter built-in in browser console (it's possible to filter content in browser console to display only "Original content", so it will be displayed only once).

**Comment by contribucious** [comment]
OK, got it. Let’s stick with the current system then. Thanks for your quick response. 😉

**Comment by AdamWr** [comment]
It's a new scriptlet which is not available yet, but if there will be more requests for similar improvements, then I think that it would be good idea to reconsider it.

**Comment by contribucious** [comment]
#### OK. 🙂 Meanwhile, can you at least consider improving the "verbose" action, to include the words "Original content:" and "Modified content:" _also_ in the `console.log` line of the object itself? _(cf. your first screenshot / [code context](https://github.com/AdguardTeam/Scriptlets/blob/6db81a29a6b6bf816005dc1703b0b7aa7d0adb53/src/scriptlets/trusted-json-set.ts#L614))_

So that when we simply search "Original content" in the browser console for example (no regex usage, etc.), that will show both its textual version and its corresponding object version? Useful to analyze all `JSON.stringify` calls of a website for example, without having everything duplicated.

_Still applying to both "Original content" and "Modified content" cases:_
Another possible solution to fix the "browser console search" system is to print both the textual and the object version in one _single_ console message (using `\n\n` for clarity before the last one, and therefore `\n\n` before the "Stack trace:" part too in this case, as there are now many parts to print in one unique message).

=> But the best is probably to keep the two existing different `console.log` lines and in fact, possibly use a clever solution: **additionally adding a specific word enclosed in parentheses specifying each version** (textual version or object version), so that we can easily filter all "Original content" but only show the desired version if wanted, or both if we want (by simply omitting the added _more specific_ word in our browser console search).

---
**Current basic working solution though:** Note that anyway, if we set a "LOGME" property (even in all `JSON.stringify` calls, _hoping the website doesn't use an anti-data-corruption detection and react to it when getting the result for some parts of the website, as content becomes altered despite the fact that we wanted only logging — i.e. a read-only mode just for inspection, initially_), searching this "LOGME" string in the browser console instead of the "Original content" or "Modified content" strings will in fact currently show it as generally wanted: both in its textual version and its object version (as this "LOGME" part is indeed present in both), in one copy only (the modified one). But the improvement described above (cf. mainly the two bold parts of this message) can still be good, I think. Don't you?

**Comment by AdamWr** [comment]
Do you mean something like on the screenshot below?

<details><summary>Screenshot</summary>

<img width="2559" height="1019" alt="Image" src="https://github.com/user-attachments/assets/bb4adfac-69ab-4a54-b02f-a748c10cad22" />

</details>

**Comment by contribucious** [comment]
Exactly! 😁
Very flexible search like so! 😊👍

**Comment by AdamWr** [comment]
Okay, I will try to add it tomorrow.
Maybe also with information what method is used, because it might be useful when two or more scriptlets are used for logging, I guess.

<details><summary>Screenshot</summary>

<img width="2559" height="1069" alt="Image" src="https://github.com/user-attachments/assets/f02fca2a-6034-4b84-802e-7030a2c1a42a" />

</details>

**Comment by contribucious** [comment]
Yes, you're right, even better!

For the format of the 3 pieces of information ("Original|Modified content", "string|object", "THE-METHOD-NAME"), we can keep it simple like that or use [], (), {}, etc. to separate them. Or use `of 'JSON.stringify'` (in this example). It's up to you, of course. 🙂

❔ By the way, which _actual_ rule did you use in your very first screenshot (cf. [your post here](https://github.com/AdguardTeam/Scriptlets/issues/308#issuecomment-4197113803)) to have an identical output in both the original and the modified content result? Not the mentioned one (apparently theoretical) as, from what I understand, in the current situation (if you used it in the scriptlet's current state), you'll have both original and modified content printed _but_ with a "LOGME" empty property appearing in modified content. Also, "log" as required condition is not found on your first screenshot.
➔ As your example is basic and the values are known in advance, maybe you simply set an existing known property to the same value just for demo purposes _(if logging is done also in this situation at least, when the value is unchanged — probably though, to at least signal it has been patched)_? And the same question for your very last screenshot (the one in your latest post).

**Comment by AdamWr** [comment]
>By the way, which actual rule did you use in your very first screenshot (cf. https://github.com/AdguardTeam/Scriptlets/issues/308#issuecomment-4197113803) to have an identical output in both the original and the modified content result?

That was something like `#%#//scriptlet('trusted-json-set', 'JSON.stringify', 'LOGME', '', 'log', '', '', '', 'true')`.
There is `log` used for `requiredInitialProps` (https://github.com/AdguardTeam/Scriptlets/blob/master/src/scriptlets/trusted-json-set.ts#L102), so it only modifies object if there is `log` property.
The stringified object was something like:
```js
JSON.stringify({"data":{"content_1":{"content":"Test"},"content_2":{"content":"Test"},"content_3":{"content":"Test"},"ads_1":{"ad":"Test"},"ads_2":{"ad":"Test"}}});
```

**Comment by contribucious** [comment]
###### Sorry in advance for the length, but I prefer to be as clear as possible — hence the clear structure too, making it easier for you to follow.<br>And ensuring we have the best scriptlet possible in AdGuard products when it becomes available for everyone. 😊<br>Take your time to read through it at your convenience during the day, like a blog post. 😉

# 1/3 — Clarification

Ah OK, so visibly the scriptlet STILL prints in the browser console DESPITE the fact that "log" is not present, but then it simply shows an identical result for both "Original content" and "Modified content". _Got it._ I logically thought it would not log at all in the specific case where the property defined in `requiredInitialProps` is not found. To be consistent with the known behavior of `json-prune` — cf. [the example in the docs](https://github.com/AdguardTeam/Scriptlets/blob/master/wiki/about-scriptlets.md#json-prune) `example.org#%#//scriptlet('json-prune', '', '"id":"117458"')` (which concerns `JSON.parse` in that one).

➔ That's great news to hear _(despite the great caveat / side effect described in the post just below though, due to that still being a workaround)_ as we therefore don't have to effectively inject a "LOGME" (or any name, for that matter) property everywhere just to have the verbose mode active (and also, as that can break some websites in rare but possible situations: websites not handling unexpected results correctly, websites using an anti-data-corruption system even on some rare parts only, etc.).

# 2/3 — Suggestion

#### By the way, can you **add the string `(changed)` _AND_ `(unchanged)`** — accordingly — to the "Modified content string" and "Modified content object" lines to quickly spot broken rules?<br>Or, alternatively, which I prefer, **replace** "Modified content" with "Unchanged content" when unchanged? (*)

➔ 🙂 Like so, if we know `example.net` has one `trusted-json-set` rule in production usage and that searching the string **`(changed)`** _(solution A)_ or the string **`Modified content`** _(solution B)_ on the correct part of the website and under the correct conditions doesn't return anything anymore in the browser console, we know it's very likely that the rule needs to be deleted or updated (typically, because the property specified in `requiredInitialProps` of the rule is no longer found nowadays — hence no trace of a change made can be found via the console).

➔ ✅ _This is excellent for easily tracking rule validity over time (to spot when a rule becomes obsolete)._

###### 💡 **(\*)** Note that for solution B, "Unchanged content" was chosen due to the fact that it's a completely different word. Unlike with "Unmodified content" where a _basic_ case-insensitive search in the console (regex not used, etc.) for "modified content" will find both "Modified content" and "Unmodified content".

_Thank you very much!_

**Comment by contribucious** [comment]
# 3/3 — The caveat

<br>

## The problem

<details>
<summary>Click to reveal …</summary>

<br>

From what I understand of your current code, it allows full logging **when the "verbose" parameter is set to true** (even in a read-only mode, so without having to set a LOGME property — for example — everywhere to show it) _but_ only in a workaround mode, meaning in the case where a given `requiredInitialProps` is not found to do so.

The best solution IMHO would be to have:
- **verbose mode:** when enabled, log only when a change has been made. Unlike now, if given `requiredInitialProps` is not present, log nothing. Useful to spot only changes done.
- **log mode:** log all, possibly limited via a limiting condition. Just like `json-prune`.
</details>

<br>

## The proposed _ideal_ solution

<details>
<summary>Click to reveal …</summary>

<br>

So, would it not be better to instead only ever log/verbose when a change has been _concretely_ made, and create instead a system _dedicated_ to logging (where possible flood of the console is allowed in any situation, no problem) like below?

### For a base, considering the `json-prune` solution:

###### ◦ Call with no arguments will log the current hostname and JSON payload at the console
`example.org#%#//scriptlet('json-prune')`

###### ◦ Call with only second argument will log the current hostname and matched json payload at the console
`example.org#%#//scriptlet('json-prune', '', '"id":"117458"')`

### Adaptation for `trusted-json-set` will logically be:

`example.org#%#//scriptlet('trusted-json-set')` ➔ For `JSON.parse` logging by default, too.
`example.org#%#//scriptlet('trusted-json-set', 'JSON.stringify')` ➔ For `JSON.stringify` logging.
`example.org#%#//scriptlet('trusted-json-set', 'JSON.stringify', '', '', 'ads.enabled')` ➔ Same, with limited logging.

###### 👀 If you think so too, this would involve creating a dedicated issue like `Improve 'trusted-json-set' — add logging support` (used for debug only, removable in production filter lists by detection of its used/unused parameters, just like `json-prune`) and letting the "verbose" action serve its initial purpose only: "verbosing" a change made.
</details>

<br>

# BONUS — Small fix idea for current situation

<details>
<summary>Click to reveal …</summary>

<br>

When we enable the verbose mode on a `trusted-json-set` scriptlet on a given website, ability to filter out console clutter via a browser console _search exclusion_ for the string `(unchanged)` / `Unchanged content` (depending on the wording chosen above).

However, ideally, that implies a small tweak to do in addition in fact, to allow a proper exclusion of the _corresponding_ "Original content string/object" messages too:  you need to have also a common specific keyword/symbol between corresponding "Original" and related "Unchanged" messages.

Or, at the very least, a unique searchable keyword/symbol to **DISTINGUISH** the 2 messages "_Original_ content string/object" related to the corresponding 2 messages "_Modified_ content string/object" (which we all want to show) **FROM** the 2 messages "_Original_ content string/object" related to the corresponding 2 messages "_Unchanged_ content string/object" (which we all don't want to show).

_It follows a clear logic, even if it adds a slight layer of complexity._
<br>

➥ To allow console filtering (inclusion or exclusion) adequately — easily, even without regex in fact —, we can imagine having something like this, for example: 👍
- `Original content string/object of JSON.stringify (of an unchanged payload)` and
- `Original content string/object of JSON.stringify (of a modified payload)`.

<hr>

<details>
<summary>Gemini suggestions / Alternatives <em>(to read only afterwards)</em></summary>

<br>

💡 **Gemini (AI) suggested wording:** payload, call or result.
Also, to use "from JSON.stringify" instead of "of JSON.stringify", to avoid using "of" twice in this case.
_"Using "from" is excellent. It clearly indicates the interception's origin—the data coming from the JSON.stringify call."_, AI explains.
➥ We can in fact use: <code>from JSON.stringify <em>call</em></code>, especially more adequate for the "_modified_ content" case: "Modified content string/object from JSON.stringify call".

_Alternative 1:_
- `Original content string/object of JSON.stringify (related|linked to an unchanged payload)` and
- `Original content string/object of JSON.stringify (related|linked to a modified payload)`.

_Alternative 2:_
- `Original content string/object of JSON.stringify (unchanged later)` and
- `Original content string/object of JSON.stringify (modified later)`.

_Alternative 3:_
- `Original content string/object of JSON.stringify (of an unchanged one)` and
- `Original content string/object of JSON.stringify (of a modified one)`.
</details>

<hr>

➥ **Searching or Excluding:**
  - "unchanged content" — finds only the "Unchanged content" log messages
  - "unchanged payload" — finds only the "Original content" corresponding log messages _(i.e. of unchanged stuff)_
  - "unchanged" — finds both

➥ **Searching or Excluding:**
  - "modified content" — finds only the "Modified content" log messages
  - "modified payload" — finds only the "Original content" corresponding log messages _(i.e. of modified stuff)_
  - "modified" — finds both

_Special alternative would be:_
- `Original content string/object of JSON.stringify (*)`
  - Related to an unchanged one.
  - Regex `/\(\*\)|Unchanged content/` for searching or excluding "Unchanged content" + its corresponding "Original content".
- `Original content string/object of JSON.stringify (**)`
  - Related to a modified one.
  - Regex `/\(\*\*\)|Modified content/` for searching or excluding "Modified content" + its corresponding "Original content".
</details>

<br>

##### UPDATE: Adjusted. Forgot that verbose state is false for production rules. Should've slept on it first!

###### _P.S. Sorry for the length. Kudos if you've read it all! And thank you on behalf of all the future users of this scriptlet!_ 😉<br>_P.P.S. [Nice stuff you've already done](https://github.com/AdguardTeam/Scriptlets/tree/fix/AG-21246-03)._ 😊

**Comment by AdamWr** [comment]
Done (I hope), except this one:
```
example.org#%#//scriptlet('trusted-json-set') ➔ For JSON.parse logging by default, too.
```

<hr>

`example.org#%#//scriptlet('trusted-json-set', 'JSON.stringify')` will log content and show it only with `Original content` message.

<details><summary>Screenshot</summary>

<img width="2559" height="812" alt="Image" src="https://github.com/user-attachments/assets/5eed27a0-3d66-41ab-a15d-3608a59faf58" />

</details>

<hr>

Regarding - `example.org#%#//scriptlet('trusted-json-set', 'JSON.stringify', '', '', 'ads.enabled')`, added support for it like in `json-prune` (string/regex searching) + support for `JSONPath`, so it should be `example.org#%#//scriptlet('trusted-json-set', 'JSON.stringify', '', '', '$.ads.enabled')` and it will log original content only if it includes specific path.

Used `#%#//scriptlet('trusted-json-set', 'JSON.stringify', '', '', '$..[/ads/]')`.
<details><summary>Screenshot</summary>

<img width="2559" height="714" alt="Image" src="https://github.com/user-attachments/assets/78ecbd0e-6b28-46ba-a3f6-97ae71d388ff" />

</details>

<hr>

In `verbose` mode, content will be printed to console only if it was modified, otherwise nothing will be logged.
And additionally added information about `requiredInitialProps` in log mode, and when `verbose` is used it will shows information about `propsPath` and `argumentValue`.

<details><summary>Screenshot</summary>

<img width="2559" height="1360" alt="Image" src="https://github.com/user-attachments/assets/f1352e34-e84a-43f1-91ab-0a1d72e846c6" />

</details>

**Comment by contribucious** [comment]
Now, _THAT_ is ultra flexible! Thank you so so much! 😊
[Code](https://github.com/AdguardTeam/Scriptlets/tree/fix/AG-21246-03) looks great! _Eager to test this!_ 👍

**Comment by contribucious** [comment]
### Consistency stuff

I guess the following **variants** might need to be updated as well once everything is 100% finalized: 😉
`trusted-json-set-fetch-response` and `trusted-json-set-xhr-response`.

And if possible, aligning `json-prune` (and its variants) as well would be great, to support advanced log filtering (via JSONPath, useful for `$.ads.enabled` for example) and also to have the similar `verbose` ability _("verbose logs only when a PRUNING/REMOVAL actually happens")_, for consistency everywhere.

### Other consistency stuff

Also, I find the parameter name `requiredInitialProps` to be better than `obligatoryProps`, as it's clearer and more explicit. The scriptlet `json-prune` now uses it in the code but the old name is still present in the docs (top of the file) **+ can still be found at some places in the file `json-prune.test.js`**.
Variants of this scriptlet (fetch/xhr) still use the old name both in the code and docs, visibly.

<details>
<summary>👀 <em>Uniformizing the description of this parameter can be nice too — <strong>click to reveal …</strong></em></summary>

<br>

➤ <code><strong>trusted-json-set</strong></code>:
```
- `requiredInitialProps` — optional, space-separated list of property paths.
  All listed paths must be present in the JSON object for the modification to occur.
```

➤ <code><strong>trusted-json-set-{fetch,xhr}-response</strong></code> variants:
```
- `requiredInitialProps` — optional, space-separated list of property paths
  which must all be present for the modification to occur.
```

➤ <code><strong>json-prune</strong></code> (and its variants) + <code><strong>evaldata-prune</strong></code>:
```
- `obligatoryProps` — optional, string of space-separated properties
  which must be all present for the pruning to occur
```
</details>

---

### Docs improvement (`trusted-json-set`)

<details>
<summary>👀 Click to reveal …</summary>

<br>

You said:
> In `verbose` mode, content will be printed to console only if it was modified, otherwise nothing will be logged.
And additionally added information about `requiredInitialProps` in log mode, and **when verbose is used it will shows information about `propsPath` and `argumentValue`**.

Maybe add a sentence in the docs for the `verbose` behavior of the logging-only mode too? Something like this for example:

<code>In logging-only mode, the original content is logged even without `verbose`.
<strong> If set to `true`, however, more technical details will be displayed.</strong></code>

Or: `Enable it if you want more technical details.`
Or: `Enable it only if you want more technical details to be displayed.`
</details>

**Comment by AdamWr** [comment]
For `trusted-json-set-fetch-response` and `trusted-json-set-xhr-response` it would be a good idea to update it later.
Regarding `json-prune` and other related scriptlets, I'm not sure if it's necessary (and worth it) to add/modify code if it can be done by using `trusted-json-set`, especially if it's just about logging.

**Comment by contribucious** [comment]
OK. 😊 By the way, I've updated my previous post with some new information. 👍

**Comment by contribucious** [comment]
Never mind regarding the `Docs improvement` part: your sentence was not very clear in fact.
The second part of your sentence was _not_ related to the first one (log mode).

In the logging-only mode, I thought you had made changes so that the `verbose` parameter, when enabled, printed _more_ technical details (which would have made sense, but we don't really have more technical details available to print in this mode in fact; and your listed ones in your sentence, `propsPath` and `argumentValue`, are indeed not intended for this logging-only mode).

I understood your logic more clearly by looking at your code in detail. 😉
